rosidl_generator_c/src/string_functions.c

```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rosidl_generator_c/string_functions.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+void
+rosidl_generator_c__String__init(rosidl_generator_c__String * str)
+{
+  str->data = NULL;
+  str->size = 0;
+  str->capacity = 0;
+}
+
+void
+rosidl_generator_c__String__fini(rosidl_generator_c__String * str)
+{
+  if (str->data) {
+    /* ensure that data and capacity values are consistent */
+    assert(str->capacity > 0);
+    free(str->data);
+    str->data = NULL;
+    str->size = 0;
+    str->capacity = 0;
+  } else {
+    /* ensure that data, szie and and capacity values are consistent */
+    assert(0 == str->size);
+    assert(0 == str->capacity);
+  }
+}
+
+bool
+rosidl_generator_c__String__assignn(
+  rosidl_generator_c__String * str, const char * value, size_t n)
+{
+  // a NULL value is not valid
+  if (!value) {
+    return false;
+  }
+  // since n + 1 bytes are being allocated n can't be the maximum value
+  if (n == SIZE_MAX) {
+    return false;
+  }
+  char * data = malloc(n + 1);
+  if (!data) {
+    return false;
+  }
```

Comment:
I would rather no add complex error handling at this level since it would complicate the generated C message code significantly. Basically it comes down to two reasons:
* wrong function arguments
* unable to mallow
If the caller ensures that it passes valid arguments the only reason is memory allocation.